File: bench/benches/polysum.exs
Code.require_file("../lib/bench.exs", __DIR__)

# map/filter/fold list pipeline: sum the squares of the odd numbers in [0, N).
# the range + intermediate lists are built INSIDE the timed work. checksum = 1333333330000.
Bench.run("polysum", fn ->
  0..19999
  |> Enum.filter(&(rem(&1, 2) == 1))
  |> Enum.map(&(&1 * &1))
  |> Enum.sum()
end)

File: bench/benches/polysum.js
// sum of squares of the odd numbers in [0, N) -- a map/filter/fold list pipeline.
// the idiomatic JS chain .filter().map().reduce() materializes the range + two
// intermediate arrays each call -- the allocation ai's deforestation fuses away.
// checksum = 1333333330000 (< 2^53, exact as a double).
const { bench } = require("../lib/bench");
bench("polysum", () =>
  Array.from({ length: 20000 }, (_, i) => i)
    .filter(x => x % 2 === 1)
    .map(x => x * x)
    .reduce((a, b) => a + b, 0));
